File: app/src/main/java/dagger/nambv/app/demo/BaseApplication.java
package dagger.nambv.app.demo;

import android.app.Application;
import android.content.Context;

import dagger.nambv.app.demo.di.component.AppComponent;
import dagger.nambv.app.demo.di.component.DaggerAppComponent;
import dagger.nambv.app.demo.di.module.AppModule;

public class BaseApplication extends Application {

    private static AppComponent appComponent;
    private static BaseApplication instance;

    @Override
    public void onCreate() {
        super.onCreate();
        instance = this;
        initAppComponents();
    }

    public AppComponent getAppComponent() {
        return appComponent;
    }

    public static BaseApplication get(Context context) {
        return (BaseApplication) context.getApplicationContext();
    }

    @SuppressWarnings("deprecation")
    private void initAppComponents() {
        appComponent = DaggerAppComponent.builder()
                .appModule(new AppModule(this))
                .build();
    }
}

File: app/src/main/java/dagger/nambv/app/demo/BaseApplication.kt
package dagger.nambv.app.demo

import android.app.Application
import android.content.Context
import dagger.nambv.app.demo.di.component.AppComponent
import dagger.nambv.app.demo.di.component.DaggerAppComponent
import dagger.nambv.app.demo.di.module.AppModule

class BaseApplication : Application() {

    override fun onCreate() {
        super.onCreate()
        instance = this
        initAppComponents()
    }

    fun getAppComponent(): AppComponent {
        return appComponent
    }

    fun initAppComponents() {
        appComponent = DaggerAppComponent.builder()
            .appModule(AppModule(this))
            .build()
    }

    companion object {

        lateinit var appComponent: AppComponent
        lateinit var instance: BaseApplication

        operator fun get(context: Context): BaseApplication {
            return context.applicationContext as BaseApplication
        }
    }
}
